creme/optim/averager.py

```diff
@@ -0,0 +1,83 @@
+import collections
+
+from . import base
+from . import schedulers
+
+
+__all__ = ['Averager']
+
+
+class Averager(base.Optimizer):
+    """An Stochastic weight averaging wrapper around various sgd optimzers.
```

Comment:
I'm not sure you've implemented the correct algorithm. I was thinking about this one: https://arxiv.org/pdf/1107.2490v2.pdf, which not the same thing. It's the one implemented by `sklearn` in it's `SGDClassifier` class.

Reply:
I will look at your proposed resources :)

I summarize what I did or atleast what I intended to do -

A `base_optim` which is the core optimizer around which we average the weights. `c` is the cycle length as I explained in above comment. The `_update_after_pred` first update the weights according to the `base_optim` and store them in `w_` for next iteration in this line,
`self.w_ = self.base_optim._update_after_pred(self.w_.copy(), g)`

Then `if not self.n_iterations % self.c` will check whether to take average (if `c` is 1 than this always evaluates to `True`) and if it's true than we update `w_swa`. Then we return `w_swa` if we averaged the weights atleast once otherwise return `w_` (the weights updated by `base_optim`).

The API may look complex because of the wrapping behaviour of `Averager`.

Also I can remove `c` to make it simple but in that case averaging will occur in every iteration.